src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt

```diff
@@ -0,0 +1,308 @@
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.model.RollupFieldMapping
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.settings.RollupSettings
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.util.getRollupJobs
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.util.populateFieldMappings
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.IndexUtils.Companion.FIELDS
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.IndexUtils.Companion.PROPERTIES
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.ActionListener
+import org.elasticsearch.action.ActionRequest
+import org.elasticsearch.action.ActionResponse
+import org.elasticsearch.action.fieldcaps.FieldCapabilities
+import org.elasticsearch.action.fieldcaps.FieldCapabilitiesRequest
+import org.elasticsearch.action.fieldcaps.FieldCapabilitiesResponse
+import org.elasticsearch.action.support.ActionFilter
+import org.elasticsearch.action.support.ActionFilterChain
+import org.elasticsearch.action.support.IndicesOptions
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.xcontent.DeprecationHandler
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.XContentFactory
+import org.elasticsearch.common.xcontent.json.JsonXContent
+import org.elasticsearch.tasks.Task
+
+private val logger = LogManager.getLogger(FieldCapsFilter::class.java)
+
+@Suppress("UNCHECKED_CAST", "SpreadOperator", "TooManyFunctions")
+class FieldCapsFilter(
+    val clusterService: ClusterService,
+    val indexNameExpressionResolver: IndexNameExpressionResolver
+) : ActionFilter {
+
+    override fun <Request : ActionRequest?, Response : ActionResponse?> apply(
+        task: Task,
+        action: String,
+        request: Request,
+        listener: ActionListener<Response>,
+        chain: ActionFilterChain<Request, Response>
+    ) {
+        if (request is FieldCapabilitiesRequest) {
+            val indices = request.indices().map { it.toString() }.toTypedArray()
+            val concreteIndices = indexNameExpressionResolver.concreteIndexNames(clusterService.state(), request.indicesOptions(), *indices)
+            val rollupIndices = mutableSetOf<String>()
+            val filteredIndices = mutableSetOf<String>()
+            for (index in concreteIndices) {
+                val isRollupIndex = RollupSettings.ROLLUP_INDEX.get(clusterService.state().metadata.index(index).settings)
+                if (isRollupIndex) {
+                    rollupIndices.add(index)
+                } else {
+                    filteredIndices.add(index)
+                }
+            }
+
+            val hasRollupIndices = rollupIndices.size > 0
+            if (hasRollupIndices) {
```

Comment:
I think if it doesn't have rollup we can just immediately call proceed right?
So something like:

```kotlin
if (rollupIndices.isEmpty()) {
    return chain.proceed(task, action, request, listener)
}
if (filteredIndices.isEmpty()) {
    val rewrittenResponse = rewriteResponse(mapOf(), arrayOf(), rollupIndices)
    return listener.onResponse(rewrittenResponse as Response)
}
request.indices(*filteredIndices.toTypedArray())
chain.proceed(task, action, request, object : ActionListener<Response> {
    override fun onResponse(response: Response) {
        logger.info("Has rollup indices will rewrite field caps response")
        response as FieldCapabilitiesResponse
        val reWrittenResponse = rewriteResponse(response.get(), response.indices, rollupIndices)
        listener.onResponse(reWrittenResponse as Response)
    }

    override fun onFailure(e: Exception) {
        listener.onFailure(e)
    }
})
```

i.e. no need to go through rest of logic as we don't need to do it.

Reply:
That looks cleaner, will update